grpc-protocol/src/main/java/com/linecorp/armeria/client/grpc/protocol/UnaryGrpcClient.java

```diff
@@ -205,32 +210,77 @@ public HttpResponse execute(ClientRequestContext ctx, HttpRequest req) {
                 private HttpData content = HttpData.empty();
                 @Nullable
                 private HttpHeaders trailers;
+                @Nullable
+                private Subscription subscription;
+                private boolean terminated;
                 private int processedMessages;
 
                 @Override
                 public void onSubscribe(Subscription s) {
+                    if (subscription != null) {
+                        s.cancel();
+                        return;
+                    }
+                    subscription = s;
                     // At least 2 requests are required for receiving trailers.
                     s.request(2);
                 }
 
                 @Override
                 public void onNext(DeframedMessage message) {
+                    try {
+                        if (terminated) {
+                            return;
+                        }
+                        process(message);
```

Comment:
it's a final consumer of the `message` and is responsible for closing it. If we do move `if` out of the `try` - it'll leak buffers. Think about this as try with resources, but I couldn't use it here because it seems like Java 8 doesn't support try with resources with references, so I had to use `finally` for closing it.

Reply:
Oops, that was a dumb comment. I misread the code. Please, forget about it. 😅

Reply:
No worries :)